Helm package for Neo4j, the world's leading graph database

mneedham opened the pull request:
Hi,

I've signed the CLA already and this is my first attempt at creating a Helm package for Neo4j. 

I'd appreciate any feedback on things I should change/improve.

Thanks, Mark

Commit:
Helm package for Neo4j, the world's leading graph database

unguiculus reviewed (changes requested):
Thanks for this PR. Please have a look at best practices, especially regarding standard labels, formatting, and values.

https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

If you create a new chart with `helm create` with Helm 2.5.+ you'll get an initial experience with current best practices applied. You may also look at the chart examples for reference.

https://github.com/kubernetes/helm/tree/master/docs/examples

Commit:
Fixing labeling

Commit:
Idiomatic values.yaml

mneedham commented:
@unguiculus cool, thanks for the feedback. 

I've read through those pages and hopefully addressed the issues with labels & values.yaml in my next two commits.

mneedham commented:
@unguiculus I'm only unsure about commenting in values.yaml. In the example there aren't comments describing each variable but the docs suggest you should describe each variable in a comment.

unguiculus commented on incubator/neo4j/templates/statefulset.yaml:
```diff
@@ -1,60 +1,62 @@
             value: CORE
           - name: NUMBER_OF_CORES
-            value: "{{default 3 .Values.NumberOfCores}}"
+            value: "{{default 3 .Values.numberOfCores}}"
```
`default 3` is unnecessary. The default value is already in `values.yaml`.

unguiculus commented on incubator/neo4j/templates/statefulset.yaml:
```diff
@@ -98,9 +100,9 @@ spec:
       name: datadir
       annotations:
-        volume.alpha.kubernetes.io/storage-class: "{{.Values.StorageClass}}"
+        volume.alpha.kubernetes.io/storage-class: "{{.Values.storageClass}}"
```
You only support alpha. You should also support beta. Here's an example for doing this:

https://github.com/kubernetes/charts/blob/master/stable/mongodb-replicaset/templates/mongodb-statefulset.yaml#L175-L199

To be honest, I wouldn't even support alpha any more.

unguiculus commented on incubator/neo4j/templates/lb.yaml:
```diff
@@ -2,20 +2,21 @@
+    app: {{ template "name" . }}
 spec:
-  type: {{.Values.ServiceType}}
+  type: {{.Values.serviceType}}
```
Please use consistent formatting everywhere: Space after `{{` and before `}}`.

unguiculus commented on incubator/neo4j/values.yaml:
```diff
@@ -3,26 +3,25 @@
+discoveryPort: 5000
+txPort: 6000
+raftPort: 7000
+resources:
```
We've started recommending not to specify defaults for resources. Also, requests and/or limits is missing.You may do it like so:

```yaml
resources: {}
# limits:
#   cpu: 100m
#   memory: 512Mi
# requests:
#   cpu: 100m
#   memory: 512Mi
```

unguiculus commented on incubator/neo4j/Chart.yaml:
```diff
@@ -0,0 +1,13 @@
+  - https://github.com/neo4j/docker-neo4j
+  - https://github.com/mneedham/k8s-kubectl
+maintainers:
+  - name: Mark Needham
```
Use Github username and add `appVersion`.

unguiculus commented on incubator/neo4j/README.md:
````diff
@@ -0,0 +1,77 @@
+## TL;DR;
+
+```bash
+$ git clone git@github.com:mneedham/charts.git
````
Wrong URL

unguiculus commented on incubator/neo4j/README.md:
```diff
@@ -0,0 +1,77 @@
+
+## Prerequisites
+
+- Kubernetes 1.4+ with Beta APIs enabled
```
1.4+: Really? We are going to support the latest two versions in the future, which are 1.6 and 1.7 at the moment.

unguiculus commented on incubator/neo4j/README.md:
````diff
@@ -0,0 +1,77 @@
+To uninstall/delete the `neo4j-helm` deployment:
+
+```bash
+$ helm delete neo4j-helm -- purge
````
Remove space `-- purge`

unguiculus commented on incubator/neo4j/README.md:
```diff
@@ -0,0 +1,77 @@
+|----------------------------|-----------------------------------------|----------------------------------------------------------|
+| `image`                    | Neo4j image                             | `neo4j`                                                  |
+| `imageTag`                 | Neo4j version                           | `{VERSION}`                                              |
+| `imagePullPolicy`          | Image pull policy                       | `Always` if `ImageTag` is `latest`, else `IfNotPresent`. |
```
`ImageTag` -> `imageTag`

unguiculus reviewed (changes requested):
Please have a look at this chart: https://github.com/kubernetes/charts/tree/master/stable/nginx-ingress/

It's a good example for a chart with multiple components. Please consider using components, adding `component` labels, and grouping values.

Commit:
Fixed naming, changed alpha -> beta, removed unnecessary defaults

Commit:
add component names

mneedham commented:
@unguiculus Have added components, removed the alpha annotation, and hopefully fixed all the spacing issues now.

unguiculus commented on incubator/neo4j/templates/core-statefulset.yaml:
```diff
@@ -0,0 +1,90 @@
+            echo "causal_clustering.raft_messages_log_enable=true" >> conf/neo4j.conf
+            echo "dbms.security.auth_enabled=false" >> conf/neo4j.conf
+
+            STATEFULSET_NAME="{{ template "fullname" . }}"
```
Should probably be `{{ template "fullname" . }}-core`

unguiculus commented on incubator/neo4j/templates/core-statefulset.yaml:
```diff
@@ -0,0 +1,90 @@
+    - metadata:
+        name: datadir
+        annotations:
+        {{- if .Values.storageClass | quote }}
```
We've just had a PR from one of the maintainers regarding the storage class. Please do it the same way.

https://github.com/kubernetes/charts/pull/1332/files

cc @mgoodness

unguiculus commented on incubator/neo4j/templates/loadbalancer.yaml:
```diff
@@ -0,0 +1,22 @@
+apiVersion: v1
```
Don't call the file loadbalancer. It's a service.

unguiculus commented on incubator/neo4j/templates/loadbalancer.yaml:
```diff
@@ -0,0 +1,22 @@
+      targetPort: {{ .Values.boltPort }}
+      name: bolt
+  selector:
+    app: {{ template "name" . }}
```
You should not make the service a component of its own but rather use the one it belongs to.

The selector needs a `component` label.

mneedham replied:
The load balancer was used for both of the other components so I wasn't sure how to label it. 

Thinking about it I don't really need it so I'll delete it.

unguiculus commented on incubator/neo4j/templates/core-dns.yaml:
```diff
@@ -0,0 +1,18 @@
+      targetPort: {{ .Values.httpPort }}
+  selector:
+    app: {{ template "name" . }}
+    release: {{ .Release.Name }}
```
Needs `component` label.

unguiculus reviewed (changes requested):
Don't overdo it with components. I've only identified two of them, core and readreplicas, if I am correct. Services should be part of the component they belong to.

Please have a look at the suggested example again. Structure values accordingly, i. e. group/nest them by component where applicable.

unguiculus commented:
Let me know if you need assistance. I can help if you like.

mneedham commented:
@unguiculus  cool, I'll do the nesting. I hadn't nested anything because of this advice in the best practices guide:

> In most cases, flat should be favored over nested. The reason for this is that it is simpler for template developers and users.

From: https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/values.md

Commit:
nest the components + don't need load balancer

mneedham commented:
I've done the changes except for the persistent storage - I'll look at that tomorrow.

Would you be free to help me finish it off during the day UK/EU time tomorrow?

unguiculus commented:
You may ping me on Kubernetes Slack, but time will be limited during the day.

Commit:
using new PVC style

unguiculus commented on incubator/neo4j/values.yaml:
```diff
@@ -0,0 +1,73 @@
+# Specs for the Neo4j docker image
+image: "neo4j"
+imageTag: "3.2.2-enterprise"
+imagePullPolicy: "Always"
```
Default should be `IfNotPresent`.

unguiculus commented on incubator/neo4j/README.md:
```diff
@@ -0,0 +1,72 @@
+## Configuration
+
+The following tables lists the configurable parameters of the Neo4j chart and their default values.
+
```
Config table needs an update.

unguiculus replied:
Please update the table.

mneedham replied:
I've updated it with the things I think users might want to use. Would you usually list every single variable in the table?

unguiculus replied:
yes

unguiculus commented on incubator/neo4j/templates/core-dns.yaml:
```diff
@@ -0,0 +1,19 @@
+    release: {{ .Release.Name | quote }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    app: {{ template "neo4j.name" . }}
+    component: "{{ .Values.core.name}}"
```
- Use consistent formatting. Space after `{{` and before `}}`.
- Don't make components configurable.

unguiculus commented on incubator/neo4j/values.yaml:
```diff
@@ -0,0 +1,73 @@
+
+# Cores
+core:
+  name: core
```
Don't make the component names configurable. Doesn't make sense IMHO.

mneedham replied:
Should I hardcode those in the yaml file or is there a way to put them in the values file and not have them configurable?

unguiculus replied:
I would hardcode them. I don't think it makes sense to make them configurable.

unguiculus commented on incubator/neo4j/values.yaml:
```diff
@@ -0,0 +1,73 @@
+httpPort: 7474
+discoveryPort: 5000
+txPort: 6000
+raftPort: 7000
```
As far as I have seen, the last three ports should go under `core`.

unguiculus commented on incubator/neo4j/Chart.yaml:
```diff
@@ -0,0 +1,13 @@
+name: neo4j
+home: https://www.neo4j.com
+version: 0.1.0
```
Add `appVersion`.

unguiculus reviewed (changes requested):
Excellent. I went over it quickly and have a few comments. Please give me some time to do a more thorough review.

unguiculus commented:
/ok-to-test

unguiculus commented:
Any reason why you require Tiller >= 2.5.1?

unguiculus commented on incubator/neo4j/templates/NOTES.txt:
```diff
@@ -0,0 +1,8 @@
+You can connect to the cluster using the Cypher shell by running:
+
+    kubectl run -it --rm cypher-shell \
```
Add `--namespace {{ .Release.Namespace }}`

unguiculus commented:
I managed to install it successfully. Sometimes, when I run the command from the NOTES.txt, it prints this message:

```
Error attaching, falling back to logs: unable to upgrade connection: container cypher-shell not found in pod cypher-shell_default
```
Otherwise it seems fine.

Complete output:

```
kubectl run -it --rm cypher-shell \                                                                                                        ± ★ ⫘ (07104c1)
        --image=neo4j:3.2.1-enterprise \
        --restart=Never \
        --command -- ./bin/cypher-shell -u neo4j -p neo4j --a donating-lamb-neo4j.default.svc.cluster.local "call dbms.cluster.overview()"
If you don't see a command prompt, try pressing enter.
                                                      Error attaching, falling back to logs: unable to upgrade connection: container cypher-shell not found in pod cypher-shell_default
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                     | addresses                                                                                                                                                                                                                                                                 | role       | groups |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| "608a41d5-d0af-40bd-82fe-82f9455d3c5f" | ["bolt://donating-lamb-neo4j-core-0.donating-lamb-neo4j.default.svc.cluster.local:7687", "http://donating-lamb-neo4j-core-0.donating-lamb-neo4j.default.svc.cluster.local:7474", "https://donating-lamb-neo4j-core-0.donating-lamb-neo4j.default.svc.cluster.local:7473"] | "LEADER"   | []     |
| "b0c13598-4844-40ca-b95b-4b354aaa29c0" | ["bolt://donating-lamb-neo4j-core-1.donating-lamb-neo4j.default.svc.cluster.local:7687", "http://donating-lamb-neo4j-core-1.donating-lamb-neo4j.default.svc.cluster.local:7474", "https://donating-lamb-neo4j-core-1.donating-lamb-neo4j.default.svc.cluster.local:7473"] | "FOLLOWER" | []     |
| "9f27e243-c976-4f2e-827f-73573ca24167" | ["bolt://donating-lamb-neo4j-core-2.donating-lamb-neo4j.default.svc.cluster.local:7687", "http://donating-lamb-neo4j-core-2.donating-lamb-neo4j.default.svc.cluster.local:7474", "https://donating-lamb-neo4j-core-2.donating-lamb-neo4j.default.svc.cluster.local:7473"] | "FOLLOWER" | []     |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

mneedham commented:
@unguiculus 

> Any reason why you require Tiller >= 2.5.1?

I took that value from the output of `helm version`. Is there a more general value I should put instead or do I just leave it out?

Commit:
more tweaks

unguiculus commented:
Just leave it out. I've never seen that specified in any chart.

Commit:
no tiller version

Commit:
hardcore component names

unguiculus commented on incubator/neo4j/templates/readreplicas-deployment.yaml:
```diff
@@ -0,0 +1,53 @@
+            value: READ_REPLICA
+          - name: NUMBER_OF_CORES
+            value: "{{ .Values.core.numberOfServers }}"
+        command:
```
I'm just thinking that it could be a good idea to actually add shell scripts as files to the chart and include them using the `Files`object. Might be easier to develop and test them. You can use a decent editor and run Shellcheck across them. I never tried this myself, but maybe you want to make a stab a this.

https://github.com/kubernetes/helm/blob/master/docs/chart_template_guide/accessing_files.md

Could also be done in a separate PR, of course.

mneedham replied:
Oooh nice idea. Let me try that as a separate PR.

unguiculus commented on incubator/neo4j/templates/tests/test-neo4j-cluster.yaml:
```diff
@@ -0,0 +1,46 @@
+    app: {{ template "neo4j.name" . }}
+  annotations:
+    "helm.sh/hook": test-success
+    "pod.beta.kubernetes.io/init-containers": '[
```
I'd suggest you use the new 1.6+ syntax for init containers.

mneedham replied:
Done

Commit:
1.6 style init containers

Commit:
adding all values in readme and parameterising everywhere

mneedham commented:
@unguiculus  table now updated.

mneedham commented:
/retest

mneedham commented:
I'm not sure why these are failing...I've tried running `helm test` locally against minikube and against GKE and it passes on both of those.

unguiculus commented:
I guess the problem is that the CI just doesn't wait long enough.

unguiculus commented on incubator/neo4j/README.md:
```diff
@@ -0,0 +1,80 @@
+|--------------------------------------|------------------------------------------------|----------------------------------------------------------|
+| `image`                              | Neo4j image                                    | `neo4j`                                                  |
+| `imageTag`                           | Neo4j version                                  | `{VERSION}`                                              |
+| `imagePullPolicy`                    | Image pull policy                              | `Always` if `imageTag` is `latest`, else `IfNotPresent`. |
```
> `Always` if `imageTag` is `latest`, else `IfNotPresent`.

Should be `IfNotPresent` only.

unguiculus commented on incubator/neo4j/values.yaml:
```diff
@@ -0,0 +1,64 @@
+
+# Read Replicas
+readReplica:
+  numberOfServers: 0
```
Why is this 0 by default?

mneedham replied:
You'd usually only add read replicas if your read load was high enough to warrant it - you don't necessarily need to have any configured by default.

Commit:
Use DNS cluster discovery (introduced in Neo 3.2.3) and hardcode the ports

mneedham commented:
I've bumped the Neo4j version to latest (3.2.3). This also lets us use DNS discovery for cluster members so it simplifies that code. 

I've also simplified things by hardcoding the ports to their defaults and removing them from the values file.

Commit:
can uncomment this part of the test too

Commit:
appVersion updated

mneedham commented:
@unguiculus Hopefully I've addressed the changes you requested. Let me know if there's anything else I should do

unguiculus commented:
Looks pretty good. May I ask you for another change? @mgoodness has come up with a great way to handle storage classes. Would be cool if you could apply that right from the start.

https://github.com/kubernetes/charts/blob/master/stable/prometheus/templates/server-pvc.yaml#L18-L30
https://github.com/kubernetes/charts/blob/master/stable/prometheus/values.yaml#L335-L377

mneedham commented:
@unguiculus yep will give that a try. I hardcoded the accessModes because Neo creates a lock file in the root directory of the database so I don't think it makes sense to allow the user to set it to something shared or read only.

rawkode commented:
Awesome. Thanks for working on this @mneedham ! Can't wait to try it out

Commit:
and back to the new style storage class

mneedham commented:
/retest

mneedham commented:
@unguiculus I added the new storage stuff. It worked for me locally but doesn't seem happy on the CI. Will retry on there

mneedham commented:
Hmmm it got much further without my change but still fails even though that one worked a few weeks ago! I tried copying all the flags locally and I was seeing an issue with the DNS service not resolving all the names of the pods which I've not seen before.

Commit:
storageClassName: standard

Commit:
empty standard class

mneedham commented:
Not sure why the CI isn't happy. I've tried it locally with minikube & on Google Cloud and it's passing for me on both of those.

unguiculus commented:
Excellent work. Thank you. Tests pass on my GKE cluster as well.
